### ISSUE TITLE
lisplambda

### ISSUE THREAD

**Issue by tromp** [woodrush/lambdalisp#5]
With BLC being such a simple language, and the LISP you implemented so feature full,
one might think that creating a lisplambda, i.e. a BLC interpreter written in your LISP dialect, should be a much easier task.

Have you ever considered doing that?

**Comment by woodrush** [comment]
I haven't thought of that yet, but it seems very interesting and feasible!
Thanks for the suggestion, I'll try writing one up!

**Comment by woodrush** [comment]
Some updates - I've finished writing up a BLC parser and a Krivine machine interpreter in LambdaLisp. I'll push this to this repo soon. I'll also try writing the byte-based I/O functions for a complete BLC interpreter.

Here's a sample output from the current interpreter:

```
$ ( cat bin/lambdalisp.blc | bin/asc2bin; cat examples/lisplambda.lisp; printf '"01010000011101000100010"' ) | bin/Blc
> 
Input: 01010000011101000100010
Lexed: (A A L L A 1 0 L 0 L 0)
Parsed: (((L L 1 0) (L . 0)) (L . 0))
----
t: (((L L 1 0) (L . 0)) (L . 0))
p: ()
e: ()
----
t: ((L L 1 0) (L . 0))
p: (((L . 0)))
e: ()
----
t: (L L 1 0)
p: (((L . 0)) ((L . 0)))
e: ()
----
t: (L 1 0)
p: (((L . 0)))
e: (((L . 0)))
----
t: (1 0)
p: ()
e: (((L . 0)) ((L . 0)))
----
t: 1
p: ((0 ((L . 0)) ((L . 0))))
e: (((L . 0)) ((L . 0)))
----
t: (L . 0)
p: ((0 ((L . 0)) ((L . 0))))
e: ()
----
t: 0
p: ()
e: ((0 ((L . 0)) ((L . 0))))
----
t: 0
p: ()
e: (((L . 0)) ((L . 0)))
----
t: (L . 0)
p: ()
e: ()
----
Result: (L . 0)
```

**Comment by tromp** [comment]
Wow; I didn't expect you to implement a Krivine machine interpreter. Is it not possible to reuse LISP's lambdas, with some extra dummy arguments to get lazy behaviour?

I once wrote this BLC interpreter in python using only its lambdas:

import sys
def c(j,t):
 sys.stdout.write(j('.')('P'))
 return t
(lambda z:lambda y:z(z(y(lambda p:lambda n:(lambda s:lambda z:z(lambda x:
lambda y:y)(lambda d:p(s)(y(s))(d)))(lambda x:lambda a:lambda s:lambda p:
p(a)(lambda y:s(n(x))(y))))(lambda c:lambda a:lambda s:z(lambda y:s(c)(y)
))))(y(lambda p:lambda b:lambda t:t(c(b,p)))))(lambda s:lambda p:p(lambda
x:lambda y:x)(s))(lambda f:(lambda q:q(q))(lambda x:f(lambda y:x(x)(y))))

I thought a similar approach could work for LISP?!

**Comment by woodrush** [comment]
That minimal BLC interpreter is very cool! I suppose it uses the universal machine proposed in your paper?

The Krivine machine code was only around 100 lines of LambdaLisp so it was pretty simple to implement. I wasn't immediately sure how to use eager Lisp lambdas for running large expressions lazily, so I stuck to the original implementation. I also thought it would make handling I/O easier but I'm not entirely sure. Resource-wise, I also believe it would be beneficial to use native lambdas if it's possible.

I put the Krivine machine interpreter here on my Gist:
https://gist.github.com/woodrush/5bc9d8e670067e06a6a3f55c14025543

One thing about handling interactive I/O is that since LambdaLisp doesn't have a byte-oriented `read`, we always must surround text inputs with `""` to feed standard input on the fly. LambdaLisp can only accept Lisp expressions from the standard input, so to input text, we must feed strings as Lisp expressions, forcing us to surround the input text with `""`. This would make restrictions for writing interactive BLC programs like games and REPLs.

Otherwise, I think I'm pretty close to having an interactive BLC interpreter, and more closer to having a static BLC interpreter with fixed input. One thing I have to figure out now is how the Krivine machine interpreter deduces which parts of the terms are in normal form during the interpretation, to print the output text interactively.

**Comment by tromp** [comment]
Oops; I see now that the python code is not an interpreter; it just implements the BLC prime number sieve:-(

I'm surprised to hear that LambdaLisp can only read Lisp expressions; isn't that rather limiting?

**Comment by woodrush** [comment]
In hindsight it sure is. At the time I was writing it I thought it would be something for a future update. Since we already have `read` I think it wouldn't be that hard to add it into the interpreter. For our purpose, I think adding a builtin that reads a string until a line feed would be sufficient.

**Comment by woodrush** [comment]
Wait, I just found that we had an undocumented function called `read-char`! I completely forgot to write it in the blog post. Here it is:
https://github.com/woodrush/lambdalisp/blob/main/src/lambdalisp.cl#L1162

Using this, I believe we can just scan the additional standard input to solve the `""` issue. I'm so glad I put that in!

**Comment by woodrush** [comment]
This was something I added to implement reader macros. We even have a built-in `peek-char` which would probably be useful too.

What we currently need is a `read-line` which reads strings from the standard input until a newline. Using `read-char` and `peek-char` I think it's straightforward to implement that as a function.

**Comment by tromp** [comment]
Yep; read-char is all you need. My IOCCC interpreter uses only the equivalent read(0,&I,1) for input.

**Comment by woodrush** [comment]
Perfect!

**Comment by woodrush** [comment]
I finished writing up the Binary Lambda Calculus interpreters! The codes are available at `examples-advanced` under the branch in #6 . I'll be merging it into the main branch after a while!

I wrote both the bit-oriented and byte-oriented interpreters. The bottleneck is the string handling, so the bit-oriented interpreter runs faster.

The great news is that I managed to run uni.blc from the IOCCC 2012 entry in about 1 or 2 minutes in this interpreter. It also runs reverse.blc from the AIT repository too.

It even runs primes.blc although it segfaults after evaluating a couple of numbers. I suspect that the segfault is probably because of LambdaLisp using up its virtual memory address space used for the variable bindings, but I'm not entirely sure.
